Improve compiler warnings visuals

viniciusmuller opened the pull request:
This pull request updates the formatting of compiler-emitted warnings, you can see some examples of the new visuals on the tests that were added.

josevalim commented on lib/elixir/src/elixir_errors.erl:
```diff
@@ -117,18 +244,30 @@ trim_line(<<$\s, Rest/binary>>, Count) -> trim_line(Rest, Count + 1);
 
+highlight_below_line(Line, Severity) ->
+  % Don't highlight leading whitespaces in line
+  {ok, Re} = re:compile("\s*", [unicode]),
```
@viniciusmuller On main, I have defined a `trim_line` function in `elixir_errors` that counts spaces. You can use it instead. :)

viniciusmuller replied:
Nice, it's much better now, thanks

josevalim commented on lib/elixir/src/elixir_errors.erl:
```diff
@@ -9,19 +9,69 @@
+    ]
+  ),
+
+  io:put_chars(standard_error, Output).
```
Sorry for the delay in reviewing. I see two concerns with this PR:

1. We are moving more logic to Erlang, which is fine, but it puts a higher barrier to entry for contribution

2. We are duplicating some logic from `format_snippet` and we should be looking for ways to unify it

Here is what I think we should do (but I am not sure):

1. Let's keep the grouping bit in Elixir, pretty much similar to how it was, but instead we call something like: `print_warning(Message, Diagnostic)`. So the grouping and the building of the message remains in Elixir.

2. We should be able to reuse format_snippet, as long as we split the file reading of the snippet from the rendering. You can think format_warning would look like this:

```erlang
format_warning(Position, File, Message, Stacktrace) ->
  Snippet = get_snippet(Position, File, Message, Stacktrace)
  format_snippet(File, LineNumber, Column, Description, Snippet).
```

Then you can make `format_snippet/4` be `format_snippet/5` (except `Snippet` is `nil`) and we can add a clause for when the column is not set either (`nil` I guess?)

viniciusmuller replied:
Thanks for your feedback!

So, if I my understanding is correct, we could:
- Reuse `format_snippet` instead of `format_line_column_diagnostic`
- Move the `format_line_diagnostic` code to another head of `format_snippet` when there's no column
- Implement the `print_warning` code when that just appends the stacktrace of grouped warnings in Elixir, but keep the formatting logic inside `format_snippet` in elixir_errors.erl
- Snippet in the *2* case would be the line, right?

josevalim replied:
Correct. You can have a single format_snippet that matches on `nil`. `print_warning` will use the given error message but everything else should come from the diagnostic.

viniciusmuller commented:
@josevalim I've updated the implementation, now all the snippet formatting logic is contained inside `format_snippet` and callers can call `get_snippet` to try and read it from a file. I've also moved the warning group printing logic to elixir inside parallel_checker.

There's just a few of those new tests failing due to probably an off-by-one error that I'll continue to debug tomorrow, but feel free to review the structural changes I've made :)

josevalim commented on lib/elixir/lib/module/parallel_checker.ex:
```diff
@@ -309,24 +309,51 @@ defmodule Module.ParallelChecker do
+    rest =
+      String.duplicate(" ", stacktrace_padding - 2) <>
+        "Invalid call also found at #{length(warn_locations)} other #{locations_plural}:\n" <>
+        :erlang.list_to_binary(warn_locations)
```
We want to make all of this part of the message, instead of printing separately, because if it is in the message, then `mix compile --all-warnings` (which is the default) will print it again with all locations.

josevalim replied:
If we make it part of the message, we don't need to call `:elixir_errors.get_line_number_digits(line, 1)` and friends either.

viniciusmuller replied:
> If we make it part of the message, we don't need to call :elixir_errors.get_line_number_digits(line, 1) and friends either.

Sorry, just saw this one now.

I made it part of the message in 32c44eddf139ece83dbe5791d7f998660475ced3 but I don't think we can avoid calling those functions from `:elixir_errors` if doing it on the elixir since, since their output is used to build the grouped warnings message

josevalim replied:
Sorry, to be clear, I think this code would be fine:


```
  defp print_warning(message, [diagnostic]) do
    :elixir_errors.print_warning(message, diagnostic)
  end

  defp print_warning(message, [diagnostic | rest]) do
    locations =
      Enum.map(rest, fn %{stacktrace: [stacktrace]} ->
        ["  ", Exception.format_stacktrace_entry(stacktrace), ?\n]
      end)

    [
      message,
      "\n\nSimilar warning found at #{length(rest)} other locations:\n",
      locations
    ]
    |> IO.iodata_to_binary()
    |> :elixir_errors.print_warning(diagnostic)
  end
```

viniciusmuller replied:
That's very nice! I thought about the formatted snippet rather than the message parameter.

I've updated the code to use this instead!

I see you changed the warning message here to `"Similar warning found at..."` rather than `"invalid call also found at..."`. 
I didn't update it here but I think it makes sense to use the warning term instead of invalid call, I can open another PR to change it if you want

josevalim replied:
Yes, let's update it in another PR. :)

josevalim reviewed (commented):
This is much better, I have added just one tiny comment!

viniciusmuller commented:
I've updated the snippet building code to emit `offset` as `column`, so that we can use a single code path for both cases by creating a `max(column - 1, 0)` spaces padding and then putting the caret symbol when highlighting a specific position. 

I think since we're the ones handling this formatting now it should be fine, as all the tests are also passing here. WDYT?

josevalim commented:
@viniciusmuller the tests are failing for me when I run them directly from root:

    elixir lib/elixir/test/elixir/kernel/diagnostics_test.exs

I think it is because some other test is defining the module `A`. Can you replace `A` by `Unknown` or `Undefined`, to avoid the chance of collisions?

josevalim commented on lib/elixir/lib/module/parallel_checker.ex:
```diff
@@ -309,24 +309,38 @@ defmodule Module.ParallelChecker do
+  defp print_warning(message, [diagnostic | rest]) do
+    total_locations = length(rest)
+
+    locations =
+      rest
+      |> Enum.with_index(1)
+      |> Enum.map(fn {%{stacktrace: [s]}, i} ->
+        if i == total_locations do
+          ["  ", Exception.format_stacktrace_entry(s)]
+        else
+          ["  ", Exception.format_stacktrace_entry(s), ?\n]
+        end
+      end)
```
```suggestion
    locations =
      Enum.map(rest, fn %{stacktrace: [s]} ->
          ["\n  ", Exception.format_stacktrace_entry(s)]
      end)
```

And then remove `\n` at the end of "Invalid call also found at #{total_locations} other #{locations_plural}:". If you want to keep it like this, you can use `Enum.intersperse` :)

josevalim commented on lib/elixir/lib/module/parallel_checker.ex:
```diff
@@ -309,24 +309,38 @@ defmodule Module.ParallelChecker do
+        end
+      end)
+
+    locations_plural = (total_locations == 1 && "location") || "locations"
```
```suggestion
    locations_plural = if total_locations == 1, do: "location", else: "locations"
```

We don't use the other style anywhere in Elixir. :)

josevalim commented:
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:
